Shader warm up

liyuqian opened the pull request:
This patch adds a default shader warm up process which moves shader compilation from the animation time to the startup time. This also provides an extension for `runApp` so developers can customize the warm up process.

This should reduce our worst_frame_rasterizer_time_millis from ~100ms to ~20-30ms for both flutter_gallery and complex_layout benchmarks. Besides, this should also have a significant improvement on 90th and 99th percentile time (50%-100% speedup in some cases, but I haven't tested them thoroughly; I'll let our device lab collect the data afterwards).

The tradeoff the is the startup time (time to first frame). Our `flutter run --profile --trace-startup` seems to be a little noisy and I see about 100ms-200ms increase in that measurement for complex_layout and flutter_gallery. Note that this only happens on the first run after install or data wipe. Later the Skia persistent cache will remove the overhead.

This also adds a cubic_bezier benchmark to test the custom shader warm up process.

This should fix https://github.com/flutter/flutter/issues/813 (either by `defaultShaderWarmUp`, or a `customShaderWarmUp`).

liyuqian committed:
[wip] Warnp Skia shaders during scheduleWarmUpFrame

liyuqian committed:
Allow custom warm up

liyuqian committed:
Improve comments

liyuqian committed:
Log timeline and fix canvas bug

liyuqian committed:
Example app to visualize defaultShaderWarmUp

liyuqian committed:
Add tests

chinmaygarde commented:
By how much does this regress the time to first frame benchmarks?

liyuqian committed:
Fix analyze

liyuqian commented:
@chinmaygarde : time to first frame increases about 100ms-200ms for complex_layout and flutter_gallery. Please see my updated PR description.

chinmaygarde commented on packages/flutter/lib/src/widgets/binding.dart:
```diff
@@ -757,6 +757,10 @@ mixin WidgetsBinding on BindingBase, SchedulerBinding, GestureBinding, RendererB
 ///
 /// Initializes the binding using [WidgetsFlutterBinding] if necessary.
 ///
+/// Provide [customShaderWarmUp] if some specific Skia shaders need to be
```
I think this can be elaborated a bit to make it more useful. Something like, "if the application has scenes that require the compilation of complex shaders, it may cause jank in the middle of an animation or interaction. Painting that scene here tells Flutter to precompute and cache the results of expensive operations so they can be reused later. This should only happen during the very first run of the Flutter application and is typically very small. It is advised that you add tracing here so that such pauses are easier to identify on the timeline. The applications main thread will not be blocked when Flutter is doing this work so you should no expect ANR warnings".

Something like that.

liyuqian replied:
Thanks! I've combined yours with my original documentation. I omit the tracing part because it's already added to `_warmUpSkiaShaderCompilations` which calls either [defaultShaderWarmUp] or [customShaderWarmUp].

CC @sfshaza2 for more API doc review.

chinmaygarde reviewed (approved):
Thanks for discussing the tradeoffs and providing data. The approach looks good to me. I am sure we can do a better job of documenting how users can use this feature but that can be tackled in future commits. LGTM but someone more familiar with dart conventions should probably also take a look.

liyuqian committed:
Move ShaderWarmUp to foundation

To avoid dependency loop

jonahwilliams commented on packages/flutter/lib/src/foundation/shader_warm_up.dart:
```diff
@@ -0,0 +1,86 @@
+/// to move common shader compilations from animation time to startup time.
+/// Alternatively, [customShaderWarmUp] can be provided to [runApp] to replace
+/// this default warm up function.
+ShaderWarmUp defaultShaderWarmUp = (Canvas canvas) {
```
You can define the function as `void defaultShaderWarmUp(Canvas canvas) {` and it will still be assignable to `ShaderWarmUp`

liyuqian replied:
I intentionally wrote `ShaderWarmUp` here to clarify that the canvas is expected to have a size of 1000x1000. (Otherwise, I could delete `ShaderWarmUp` and just use the signature directly everywhere.)

jonahwilliams replied:
I mean in terms of how the function is declared:

```suggestion
void defaultShaderWarmUp(Canvas canvas) {
```

liyuqian replied:
I understood that. My intention is that when developers see the signature of `defaultShaderWarmUp`, they can discover a link to `ShaderWarmUp` and realize that we're doing work on a 1000x1000 canvas. Does that help?

jonahwilliams replied:
If you leave a comment in the dartdoc `[ShaderWarmup]` they can follow that link as welll

liyuqian replied:
Done. Thanks!

jonahwilliams commented on packages/flutter/lib/src/foundation/shader_warm_up.dart:
```diff
@@ -0,0 +1,86 @@
+  ];
+
+  // Warm up path stroke and fill shaders.
+  for (Path path in paths) {
```
If we're trying to make this as fast as possible, index iteration is slightly faster than an iterator protocol

liyuqian replied:
Done.

Hixie replied:
might want to benchmark this. I'm not sure how true that really is in practice.

liyuqian committed:
No Colors in foundation

liyuqian committed:
Improve doc and loop

goderbauer commented:
@liyuqian Is this ready to land?

liyuqian commented:
@goderbauer I guess so if the review from @jonahwilliams is a green light on the Dart side.

BTW, as I'm digging emails, I see that we're in a tree freeze due to P0 Google3/Fuchsia roll failure. Should I wait the tree to be reopened before merging this?

jonahwilliams commented:
per the addition to runApp, I do not have a strong opinion on the API but others might

jonahwilliams reviewed (approved):
LGTM with nit

liyuqian commented:
CC @Hixie for the review of API change to `runApp`.

liyuqian committed:
Simplify defaultShaderWarmUp signature

Hixie commented on packages/flutter/lib/src/foundation/shader_warm_up.dart:
```diff
@@ -0,0 +1,84 @@
+import 'dart:ui';
```
please see the README in this directory.

This probably belongs in the `painting` library, not `foundation`.

liyuqian replied:
Moved to painting. Painting was my first choice but analyze won't allow me to do that since I used it in scheduler, which creates a dependency loop. Now I've removed all changes to scheduler and `ShaderWarmUp` is in painting again.

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -5,10 +5,11 @@
 import 'dart:collection';
 import 'dart:developer';
-import 'dart:ui' show AppLifecycleState;
+import 'dart:ui' show AppLifecycleState, Canvas, Image, Picture, PictureRecorder;
```
the scheduler shouldn't know about these things. They have nothing to do with scheduling.

liyuqian replied:
Removed changes to scheduler.

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -718,12 +719,33 @@ mixin SchedulerBinding on BindingBase, ServicesBinding {
+    if (customShaderWarmUp != null) {
+      customShaderWarmUp(canvas);
+    } else {
+      defaultShaderWarmUp(canvas);
```
You could just make `defaultShaderWarmUp` the default value of `customShaderWarmUp`.

liyuqian replied:
I discussed with @zanderso and it seems that Dart doesn't support default value of non-const variable (`defaultShaderWarmUp` is not const).

Hixie replied:
This code prints "foo":

```dart
void foo() { print('foo'); }

void bar([ void	Function() arg = foo ]) { arg(); }
      
void main() {
  bar();
}
```

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -718,12 +719,33 @@ mixin SchedulerBinding on BindingBase, ServicesBinding {
+    final Picture picture = recorder.endRecording();
+    final TimelineTask shaderWarmUpTask = TimelineTask();
+    shaderWarmUpTask.start('Warm-up shader');
+    picture.toImage(1000, 1000).then((Image image){
```
trivial nit: missing space before `{`

Hixie commented on packages/flutter/lib/src/scheduler/binding.dart:
```diff
@@ -718,12 +719,33 @@ mixin SchedulerBinding on BindingBase, ServicesBinding {
+    final TimelineTask shaderWarmUpTask = TimelineTask();
+    shaderWarmUpTask.start('Warm-up shader');
+    picture.toImage(1000, 1000).then((Image image){
+      shaderWarmUpTask.finish();
```
this doesn't finish the shaderWarmUpTask, it finishes the last task that was started. Since we don't guarantee ordering here I think that means that we're not going to get coherent results in practice.

Hixie replied:
nevermind, i misread the docs.

Hixie commented:
Doing this on startup seems weird, since we're about to render the whole scene anyway. Don't we actually want to do this _after_ startup?

liyuqian committed:
Nits

liyuqian committed:
Default parameter

liyuqian committed:
Move ShaderWarmUp to painting

liyuqian reviewed (commented):
Regarding warm up after startup: what if the app startup is an animation so we could see a jank in that animation?

Independently, Chinmay and I are planning to move the picture image snapshot (not the layer tree snapshot) to the IO thread. That would allow this warm up and the first frame to be rendered simultaneously.

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    // Warm up shadow shaders.
+    const Color black = Color(0xFF000000);  // foundation can't access material/colors
```
this comment seems weird. Why would we need to access material/colors? And why would it matter whether foundation can? (I assume this just wasn't updated when you moved it.)

Hixie replied:
(looks like you missed this comment)

Hixie commented:
Regarding where to call the warm-up: I guess the documentation in this patch doesn't really tell me enough to really evaluate things enough to review the patch.

I think the documentation should cover all of the following at a minimum, probably many more things that I'll think about once I know the answers to these questions. :-)

- Does the warm-up happen synchronously relative to rendering frames, or will the frames keep rendering while the warm-up happens in the background? If it's asynchronous, when does it happen? Does it use CPU/GPU that slows down the frame rendering, does it happen while the threads are idle, something else?
- Why does the warm-up have to happen every time the framework is started, instead of (say) when the engine is compiled, or when the app is compiled, or when the app is first run on the device?
- How do you know what to write in a shader warm-up function?
- What happens if you don't warm up the shaders? e.g. if the runApp triggers an animation, will that animation already force all the relevant shaders to be warmed up, negating the need for a separate warm-up?
- Why 1000x1000? Why not 512x512 or 10x10 or 4096x4096? Or 5x1000? How do you know what size to make it?
- Do things vary from device to device?
- Some of the things we warm up seem really obvious and simple, like "stroke shaders". Why is this something that should be done from Dart code? Won't it affect _any_ framework that uses dart:ui? Should Skia not do this itself?
- We warm up text shaders. Are those font-specific? We use the string "Hello, World". Why not the empty string, or a space character, or "مرحبا"?
- Does any of this matter with the Web backend?

liyuqian committed:
More docs

liyuqian commented:
> Regarding where to call the warm-up: I guess the documentation in this patch doesn't really tell me enough to really evaluate things enough to review the patch.
> 
> I think the documentation should cover all of the following at a minimum, probably many more things that I'll think about once I know the answers to these questions. :-)
> 
> * Does the warm-up happen synchronously relative to rendering frames, or will the frames keep rendering while the warm-up happens in the background? If it's asynchronous, when does it happen? Does it use CPU/GPU that slows down the frame rendering, does it happen while the threads are idle, something else?

Currently the warm-up happen synchronously on the GPU thread which means the rendering of the first frame on the GPU thread will be postponed until the warm-up is finished. *I’ve added this to the comment.*
> * Why does the warm-up have to happen every time the framework is started, instead of (say) when the engine is compiled, or when the app is compiled, or when the app is first run on the device?

The warm-up depends on the device’s GPU and driver so it can’t be done during compilation. The warm-up indeed only compiles the shader during the first run on the device. The current approach will cause a small overhead of drawing (without shader compilations) in subsequent runs. See the doc in https://github.com/flutter/flutter/pull/27660/files#diff-e158ae61bb7a8dd81f710e8f59aa23caR764.
> * How do you know what to write in a shader warm-up function?

*I’ve added the following to the doc.*

The draw operations being warmed up here are decided according to Flutter engineers' observation and experience based on the apps and the performance issues seen so far.

To decide which draw operations to be added to your custom warm up process, try capture an skp using `flutter screenshot --observatory- port=<port> --type=skia` and analyze it with https://debugger.skia.org. Alternatively, one may run the app with `flutter run --trace-skia` and then examine the GPU thread in the observatory timeline to see which Skia draw operations are commonly used, and which shader compilations are causing janks.

> * What happens if you don't warm up the shaders? e.g. if the runApp triggers an animation, will that animation already force all the relevant shaders to be warmed up, negating the need for a separate warm-up?

Suppose that there’s another animation later that’s very different from the start up animation. Without the shader warm-up of draw operations for the later animation, it will jank. With the warm-up, there’s a startup latency before the start up animation, and there will be no later janks.

> * Why 1000x1000? Why not 512x512 or 10x10 or 4096x4096? Or 5x1000? How do you know what size to make it?

*Added the following to the doc*
The size of [width]=[height]=1000 here are picked arbitrarily. The only goal is to have them not too far away from the actual resolution (very few mobile devices seem to have a resolution that is 10x larger or 10x smaller than 1000x1000). The size should have little effect on the shader warm-up. A custom shader warm up can pick another arbitrary number that's close to the targeted devices.
> * Do things vary from device to device?
Yes for the compiled shaders. *Added the following to the doc:*

This warm up needs to be run on each individual device because the shader compilation depends on the specific GPU hardware and driver a device has. It can't be pre-computed during the Flutter engine compilation as the engine is device agnostic.
> * Some of the things we warm up seem really obvious and simple, like "stroke shaders". Why is this something that should be done from Dart code? Won't it affect _any_ framework that uses dart:ui? Should Skia not do this itself?

I guess the key question here is who has the information about when to initiate the warm-up process. Both dart:ui and Skia probably don’t know when is a good time to do a warm-up. The framework has such information (it knows when the app is first started). In the end, the warm-up is just calling some Skia draw functions directly. It’s probably not too different from Skia providing a warm-up function API which is then called by the framework at the right time.

> * We warm up text shaders. Are those font-specific? We use the string "Hello, World". Why not the empty string, or a space character, or "مرحبا"?

I didn’t know that we already warmed up text shaders. In that case, I can just delete the “Hello, World”.

> * Does any of this matter with the Web backend?

It probably depends on how Flutter for Web is implemented and how the web browser (mostly Chromium?) calls Skia. I’ll ask yegor and the Skia team.

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+import 'dart:developer';
```
missing license

Hixie replied:
consider adding a test that verifies that all source files have a license

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+import 'dart:developer';
+import 'dart:ui';
```
`as ui`

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+/// When Skia first sees a certain type of draw operations on GPU, it needs to
+/// compile the corresponding shader. The compilation can be slow (20ms-200ms).
+/// Having that time as a startup latency is much better than having a jank in
+/// the middle of an animation.
```
> Having that time as a startup latency is much better than having a jank in the middle of an animation.

I think we should say it's "often" better, and make sure we provide a way for people to opt-out.

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+/// Having that time as a startup latency is much better than having a jank in
+/// the middle of an animation.
+///
+/// Therefore we use this during the [runApp] call to move common shader
```
comma after "Therefore"

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+  int get width;
+
+  /// Height of the image to be drawn.
+  int get height;
```
rather than width and height, return a Size.

Also, consider adding commentary explaining what size to return (e.g. "The size should be similar to the size of the device's screen. 1024x1024 is a good choice as it is within an order of magnitude of most devices.").

Why don't we use the actual screen size? I guess we don't know it early enough in release builds?

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+
+  /// Trigger draw operations on a given canvas to warm up GPU shader
+  /// compilation cache.
+  void warmUpOnCanvas(Canvas canvas);
```
this should be `@protected`

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+/// engineers' observation and experience based on the apps and the performance
+/// issues seen so far.
+///
+/// To decide which draw operations to be added to your custom warm up process,
```
I would move this paragraph to the documentation of `warmUpOnCanvas` above.

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+  int get width => 1000;
+
+  @override
+  int get height => 1000;
```
if we're going to pick an arbitrary number, 1024 is probably more idiomatic...

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+  /// compilation cache.
+  @override
+  void warmUpOnCanvas(Canvas canvas) {
+    final RRect rrect = RRect.fromLTRBXY(20, 20, 60, 60, 10, 10);
```
please use `.0` on double literals so it's clear that they're doubles. also applies in other places below.

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+    );
+
+    // The following path is copied from
+    // https://skia.org/user/api/SkCanvas_Reference#SkCanvas_drawPath
```
s/copied/based on the path/

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+    ];
+
+    // Warm up path stroke and fill shaders.
+    for (int i = 0; i < paths.length; i++) {
```
fwiw my preference is `i += 1` rather than `i++`, it avoids the confusion about whether the compiler is going to know to treat it as `++i` and discard the value it would otherwise keep around redundantly.

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,144 @@
+    canvas.translate(0, 80);
+    final ParagraphBuilder paragraphBuilder = ParagraphBuilder(
+      ParagraphStyle(textDirection: TextDirection.ltr),
+    )..pushStyle(TextStyle(color: black))..addText('Hello, world.');
```
let's pick some more minimal text like the empty string or a space or an emoji or something that isn't quite so arbitrary as "Hello world".

Hixie commented on packages/flutter/lib/src/widgets/binding.dart:
```diff
@@ -765,10 +778,13 @@ mixin WidgetsBinding on BindingBase, SchedulerBinding, GestureBinding, RendererB
     ..attachRootWidget(app)
     ..scheduleWarmUpFrame();
+  shaderWarmUp.execute();
 }
```
I think now that we've moved this to the painting library, it's more clear that `runApp` is the wrong layer to call this from. I'd call this from the painting library's binding's initialization logic. Maybe have a getter there that returns the ShaderWarmUp instance to use.

You can still take a `shaderWarmUp` method here, just pass it to `WidgetsFlutterBinding.ensureInitialized` and have the WidgetsBinding implementation override the painting library's getter to return the provided value if it's not null.

liyuqian committed:
Improve docs and replace w/h with size

liyuqian committed:
Move custom ShaderWarmUp to PaintingBinding

liyuqian committed:
Import foundation.dart

liyuqian committed:
Resolve circular dependency

Hixie commented on dev/benchmarks/macrobenchmarks/lib/src/cubic_bezier.dart:
```diff
@@ -0,0 +1,376 @@
+import 'dart:math';
```
missing license

please consider writing an automated test that verifies that all source files have a license header

liyuqian replied:
Done. Will add a test in another PR.

Hixie commented on dev/benchmarks/macrobenchmarks/lib/common.dart:
```diff
@@ -1 +1,2 @@
 const String kCullOpacityRouteName = '/cull_opacity';
```
missing license

liyuqian replied:
Done.

Hixie commented on dev/benchmarks/macrobenchmarks/lib/main.dart:
```diff
@@ -1,6 +1,7 @@
 import 'package:flutter/material.dart';
```
missing license

liyuqian replied:
Done.

Hixie commented on dev/benchmarks/macrobenchmarks/test_driver/util.dart:
```diff
@@ -0,0 +1,40 @@
+import 'dart:async';
```
missing license

liyuqian replied:
Done.

Hixie commented on packages/flutter/lib/src/painting/shader_warm_up.dart:
```diff
@@ -0,0 +1,148 @@
+  /// order of magnitude of most devices.
+  ///
+  /// A custom shader warm up can override this based on targeted devices.
+  ui.Size get size => const ui.Size(1024, 1024);
```
please use `.0` on integral double literals

liyuqian replied:
Done.

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

cc @goderbauer

liyuqian committed:
Adding missing licenses.

liyuqian commented:
This patch adds a default shader warm up process which moves shader compilation from the animation time to the startup time. This also provides an extension for `runApp` so developers can customize the warm up process.

This should reduce our worst_frame_rasterizer_time_millis from ~100ms to ~20-30ms for both flutter_gallery and complex_layout benchmarks. Besides, this should also have a significant improvement on 90th and 99th percentile time (50%-100% speedup in some cases, but I haven't tested them thoroughly; I'll let our device lab collect the data afterwards).

The tradeoff the is the startup time (time to first frame). Our `flutter run --profile --trace-startup` seems to be a little noisy and I see about 100ms-200ms increase in that measurement for complex_layout and flutter_gallery. Note that this only happens on the first run after install or data wipe. Later the Skia persistent cache will remove the overhead.

This also adds a cubic_bezier benchmark to test the custom shader warm up process.

This should fix https://github.com/flutter/flutter/issues/813 (either by `defaultShaderWarmUp`, or a `customShaderWarmUp`).

i-schuetz commented:
I saw some back and forth with this PR in master, is now the final version merged?
Asking because I'm running the gallery demo from master on my iPhone and it's still janky the first time. Up to date, with `flutter run --release`.

liyuqian commented:
I assume that it's still janky when you reopen the app and run the animation for the first time? If so, I guess that it's caused by something else. Can you please run the gallery demo with `flutter run --profile --trace-skia`, save the trace from observatory timeline, and upload it here so we can take a deeper look?

i-schuetz commented:
The same happens after closing the app (where closing means really closing, not only sending to the background).

Attached the trace. This is of a new installation, where I opened "Studies". Then for each of the 4 items inside, opened it and pressed back. There was jank in all of them (it skips a few frames, such that the push animation almost jumps to its finished state, in some more than the others).

[dart-timeline-2019-2-19.json.zip](https://github.com/flutter/flutter/files/2982066/dart-timeline-2019-2-19.json.zip)

msw333 commented:
I can confirm that jittering on first startup still exists on iOS (just create any type of app with a list with 30 items with listview.builder and scroll it when you launch the app).

Hixie commented:
I recommend that anyone running into slowness on startup file a bug with a small repro test case. There are probably many unrelated causes. This PR was to solve #813 which was specifically about the slowness in the Gallery.

i-schuetz commented:
@Hixie My comment was about the gallery (not sure if you're addressing me as well).

liyuqian commented:
I think Hixie's comment is only for @msw333 . For that list with 30 items issue, it would helpful if @msw333 could create a new issue, attach the trace and a single-file Flutter app that can allow us to reproduce the issue.

For @i-schuetz 's issue, yes it's currently still expected to have a little jank right after the new installation. Your trace is consistent with my local experiment with a new installation. But once I close the app and reopen it on my iPhone6S, the jank would disappear if I open the same item. Can you please confirm that it's also the case in your iPhone6? If not, can you please attach the trace of the second first-run after the installation?

Hixie commented:
We care about first-run performance too, very much in fact. :-)

msw333 commented:
Ok ;-). Hope that helps. Issue is observable in every flutter app I was using. iPhone X, 12.1.4 (iOS). 
Issue is more observable when using a list which is a little more complex.

**Videos don't give it justice, but show the issue:**
https://www.youtube.com/watch?v=oHKNKi08qw8
https://www.youtube.com/watch?v=1mdM_Uv574w

**Code to reproduce (just use as main.dart). Again, I can reproduce it with any flutter app on iOS. Especially observable within lists:**
https://gist.github.com/msw333/dd0a06fb1fe8b4eb4ae627542202d43a

liyuqian commented:
Thanks @msw333 ! That's very helpful. Let's further discuss this issue on https://github.com/flutter/flutter/issues/28113

i-schuetz commented:
Attached the trace from opening the 4 items after closing the app.

Also attached videos, one for directly after launching and another for after closing. As @msw333 says they don't do it justice, but still clearly noticeable.

As mentioned before after closing is better, but there's still jank. 

[trace_after_close.zip](https://github.com/flutter/flutter/files/3007199/trace_after_close.zip)
[videos.zip](https://github.com/flutter/flutter/files/3007204/videos.zip)
